### DISCUSSION TITLE
[next/image] Responsive images with "art direction"

### DISCUSSION THREAD

**Discussion by jasonsilberman** [vercel/next.js#19880] [Help] [answered]
I am using the `Image` component and it is working quite well, however, I was wondering if it is possible to implement "art direction" as described in the [MDN docs here](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#Art_direction)?

The idea is that based on media queries (all I need is a certain width breakpoint in this case), change the src of the image. For example, on desktop have a more landscape image, and on mobile show an image that is more of a square.

Here is the example from the [MDN docs](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#Art_direction):

```html
<picture>
  <source media="(max-width: 799px)" srcset="elva-480w-close-portrait.jpg">
  <source media="(min-width: 800px)" srcset="elva-800w.jpg">
  <img src="elva-800w.jpg" alt="Chris standing up holding his daughter Elva">
</picture>
```

How can I achieve this using the Image component? My only idea so far is to have multiple `Image`s and use a wrapper component that shows/hides based on a css media query. I think this would work, however, this seems less than ideal since it will probably lead to multiple or incorrect image loads.

Thanks.

**Comment by g-elwell** [comment]
Your suggestion should work fine. If you hide your elements by using CSS `display: none` then the unused images won't be loaded by the browser.

**Reply by Sandstedt** [reply]
Just tried this solution, and even if it's a bit ugly, it seems to work fine. Have tried in latest Chrome and Firefox at least. Here's my implementation inside a hero component with CSS `position: relative` and `height: 100vh`:

```js
<style>
{`
  .hero__desktop-image {
    display: none !important;
  }
  @media(orientation: landscape) {
    .hero__desktop-image {
      display: block !important;
    }
    .hero__mobile-image {
      display: none !important;
    }
  }
`}
</style>
<Image
  className="hero__mobile-image"
  src={sourceToPortraitImage}
  sizes="100vw"
  alt={image.alt}
  layout="fill"
  objectFit="cover" 
/>
<Image
  className="hero__desktop-image"
  src={sourceToLandscapeImage}
  sizes="100vw"
  alt={image.alt}
  layout="fill"
  objectFit="cover"
/>
```

_Note the !important, due to next/image use of inline-style that set's a display: block per default._

**Reply by Sandstedt** [reply]
But with that said. I must say it's much easier to debug what image being loaded in the browser using this <img> method instead of the suggested <picture> with multiple <source> elements solution. In the current way, you can just look at the src in your dev tool, and see what image size is being served, to see if your sizes string is correct.

Also interesting, when setting display: none, using my code above, on an image using next/image, next.js will serve a 1x1 base64 encoded pixel instead. Smart!

**Reply by dfellows12** [reply]
Thanks for the help!

**Reply by aNyMoRe0505** [reply]
note: this work because next/image default is [lazy load](https://nextjs.org/docs/api-reference/next/image#loading)

**Reply by thexpand** [reply]
@aNyMoRe0505 Absolutely! When I try to use `priority={true}`, sadly, it no longer works.

**Reply by rein96** [reply]
thanks for sharing!

I tried this one, and it works fine for me (with Tailwind)

```tsx
         <Image
            alt={image.name}
            className="w-full hidden md:block"
            height={0}
            sizes="100vw"
            src={image?.asset?.url}
            width={0}
          />
          <Image
            alt={mobileImage.name}
            className="w-full block md:hidden"
            height={0}
            sizes="100vw"
            src={mobileImage?.asset?.url}
            width={0}
          />
```

**Reply by bobaaaaa** [reply]
We also did this. The downside is, that both images are requested by the browser. So you always end up downloading to much

**Comment by grahammather** [comment]
The problem with this approach is when you set `priority={true}`  on your Image. Both the mobile and the desktop images will be preloaded because there's no media query in the preload link. We need to be able to do [Responsive Preloading](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload#including_media) in order to avoid preloading the desktop image on mobile. AFAIK, this isn't possible right now.

**Reply by aghoneim92** [reply]
I'd like to open an issue for this

**Reply by aghoneim92** [reply]
https://github.com/vercel/next.js/issues/29612

**Reply by gkpty** [reply]
responsive preloading would be an awesome addition to the next image component.. right now im having performance issues because my header image is different for web and mobile and im preloading both

**Comment by BL1133** [comment]
I've had this issue before. The problem with display: none is I believe the images are still downloaded, so it's an insane cost to have like 3 downloads for every image at all times. So I ended up making up all images load lazy so that this doesn't happen. It's not a great work around at all because lazy loading every image isn't optimized either, but it wasn't a production site. I think the only way to do it correctly is with javascript and not CSS media queries

**Comment by Multiply** [comment]
We ended up writing our own `<NextPicture />` component, which takes some of the capabilities of next/image, but with support for loading different sources per breakpoint.
It's was extremely difficult landing on a good React API for this feature, as you can do so many different things with sources, media queries etc.

We'll iterate some more internally, and publish it at a later time, when we've tested it fully.

**Reply by kronaemmanuel** [reply]
Did you end up publishing the `<NextPicture />` component?

**Reply by Multiply** [reply]
Our implementation is very specific to our use-case still. Essentially it only supports switching source image on various breakpoints at the moment.

It's hardly worth publishing in its current state.

**Reply by thexpand** [reply]
@Multiply Actually, you'd be surprised how many people would be thrilled to be able to do just that - change sources on certain breakpoints. My team and I have had numerous cases in which changing the source image would've saved us. It would be awesome if you could at least share a gist of your generic wrapper component for `next/image`.

**Reply by ncorso** [reply]
Completely agree with @thexpand! It would be very kind of you @Multiply if you can share your current implementation.
At least to have as a starting point :)

**Reply by AdrianFahrbach** [reply]
Do you have an Update on this? Would be interesting to see your approach!

**Reply by drMonro** [reply]
Need update on this!!

**Reply by MartinaMilisic** [reply]
Would love to see your solution on this :)

**Reply by bobaaaaa** [reply]
FYI: https://github.com/humaans/next-img has `<picture>` support with art-direction

**Comment by leerob** [comment, marked as answer]
**Update:** We have extracted the core logic from `next/image` into a new `unstable_getImgProps()` function.

This allows usage outside of `<Image>`, such as:

1. Working with [`background-image`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-image) or [`image-set`](https://developer.mozilla.org/en-US/docs/Web/CSS/image/image-set)
2. Working with canvas [`context.drawImage()`](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Using_images) or simply `new Image()`
3. Working with [`<picture>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture) media queries to implement Art Direction or Light/Dark Mode images

### Example

```tsx
import { unstable_getImgProps as getImgProps } from 'next/image'

export default function Page() {
  const common = { alt: 'Hero', width: 800, height: 400 }
  const { props: { srcSet: dark } } = getImgProps({ ...common, src: '/dark.png' })
  const { props: { srcSet: light, ...rest } } = getImgProps({ ...common, src: '/light.png' })

  return (<picture>
  <source media="(prefers-color-scheme: dark)" srcSet={dark} />
  <source media="(prefers-color-scheme: light)" srcSet={light} />
  <img {...rest} />
</picture>)
}
```

PR: https://github.com/vercel/next.js/pull/51205

**Reply by eroSilva** [reply]
Amazing, @leerob!

Just a question... why do you set the function name with prefix `unstable_`? This is an vercel process to enable beta features?

**Reply by bobaaaaa** [reply]
@eroSilva It's just not ready for prod. Use it at your own risk. React does the same:
https://github.com/facebook/react/blob/546fe4681c52de6a333a55cedb141c87b626425e/packages/react/src/React.js#L141

**Reply by AlexandreBourdeaudhui** [reply]
Hello @leerob 👋 , do you know why when we're using the `images.unoptimized: true` in `next.config.js`, the srcset output into `<source>` are empty ?


⏬  Optimized

<img width="1617" alt="Capture d’écran 2024-03-30 à 16 37 39" src="https://github.com/vercel/next.js/assets/24500083/fd02affe-6f9b-4a2c-acc6-358af3c702c5" />

⏬  Unoptimized
<img width="1695" alt="Capture d’écran 2024-03-30 à 16 37 21" src="https://github.com/vercel/next.js/assets/24500083/d00362b1-a730-4f81-a66a-19c87faed06d" />

**Reply by coelhucas** [reply]
To anyone getting here, in Next.JS 14.X we have it working as `getImageProps` 🙂

**Reply by NickWoodward** [reply]
Hi Rob, hope you don't mind me asking a noob question - where is `rest` coming from here? ie how can I set the default image? 
Thanks :)

**Reply by revnelson** [reply]
> Hi Rob, hope you don't mind me asking a noob question - where is `rest` coming from here? ie how can I set the default image? Thanks :)

In the example code from @leerob's post you can see:

```
  const { props: { srcSet: dark } } = getImgProps({ ...common, src: '/dark.png' })
  const { props: { srcSet: light, ...rest } } = getImgProps({ ...common, src: '/light.png' })
```

`rest` there is the spread object of all of `getImgProps` returned values except for `srcSet`.

**Comment by cvbuelow** [comment]
Art Direction hero image with preloading:
```jsx
import {
  unstable_getImgProps as getImgProps,
  ImageProps,
  StaticImageData,
} from 'next/image'
import Head from 'next/head'

const AdaptiveImage = ({
  className,
  alt = 'Hero image',
  fill = true,
  priority = true,
  sizes = '100vw',
  breakpoint = 640,
  desktopImage,
  mobileImage,
  style = { objectFit: 'cover' },
  ...props
}: {
  breakpoint?: number
  desktopImage: StaticImageData
  mobileImage: StaticImageData
} & Partial<ImageProps>) => {
  const commonPreload = {
    rel: 'preload',
    as: 'image',
    imageSizes: sizes,
  }
  const common = { alt, fill, priority, sizes, style, ...props }
  const { srcSet: desktop } = getImgProps({ ...common, src: desktopImage }).props
  const { srcSet: mobile, ...rest } = getImgProps({ ...common, src: mobileImage }).props
  const desktopMedia = `(min-width: ${breakpoint}px)`
  const mobileMedia = `(max-width: ${breakpoint - 1}px)`
  return (
    <>
      <Head>
        <link
          {...commonPreload}
          media={desktopMedia}
          href={desktopImage.src}
          imageSrcSet={desktop}
        />
        <link
          {...commonPreload}
          media={mobileMedia}
          href={mobileImage.src}
          imageSrcSet={mobile}
        />
      </Head>
      <picture className={className}>
        <source media={desktopMedia} srcSet={desktop} />
        <source media={mobileMedia} srcSet={mobile} />
        <img alt={alt} {...rest} />
      </picture>
    </>
  )
}

export default AdaptiveImage

```
Usage:
```jsx
import HeroDesktop from './hero-desktop.png'
import HeroMobile from './hero-mobile.png'

<AdaptiveImage
  alt='My Hero'
  desktopImage={HeroDesktop}
  mobileImage={HeroMobile}
/>
```

**Reply by ddaoxuan** [reply]
Isn't that downloading both images no matter the view anyway ?

**Reply by cvbuelow** [reply]
no, it only downloads one, either desktop or mobile image depending on screen size. if you resize your screen across the breakpoint it will then download the other one. you can see it in action here https://thangs.com/memberships/discover (on the hero image)

**Comment by Emanuele-Fortunati** [comment]
It's a great solution, unfortunately it won't work with the App Router. Any idea on how to make it working?

**Reply by Emanuele-Fortunati** [reply]
I've actually found the solution myself: just add 

`        <link
          {...commonPreload}
          media={desktopMedia}
          href={desktopImage.src}
          imageSrcSet={desktop}
        />
        <link
          {...commonPreload}
          media={mobileMedia}
          href={mobileImage.src}
          imageSrcSet={mobile}
        />`

inside the <picture> tag and Next will move it to the head tag

**Reply by revnelson** [reply]
It seems some of your response was cut off. Would you mind helping me (and others) with your solution for the app router?

**Reply by Emanuele-Fortunati** [reply]
Sure. You just need to add that code snippt anywhere in your component. The NextJs engine will copy/paste it into the <head> tag itself.

**Reply by revnelson** [reply]
Thank you!

**Reply by Arunjenson** [reply]
@Emanuele-Fortunati im using the same solution provided above in 14.x version ,but its not working, is it possible to share the whole component so that it will be clear for us?

**Reply by revnelson** [reply]
@Arunjenson Here is a gist to the component as I use it. You will need to tweak it to work for your use case, but hopefully it helps!

https://gist.github.com/revnelson/046b87cfb981be0909bc9e475efacdef

**Comment by lubojanski** [comment]
to anyone looking for solution to this problem in 2025, the getImageProps api is stable now,
link includes example.
https://nextjs.org/docs/app/api-reference/components/image#getimageprops

**Comment by kncalderong** [comment]
Another option I found useful and simpler is creating a component using three Image blocks (from Nextjs). It still leverages all the performance optimizations from Nextjs component but conditionally loads one image source or another. This could be an starting point as more configs can be added
```js
import Image from "next/image";

type PictureProps = {
  mobile: string;
  tablet: string;
  desktop: string;
  alt: string;
};

export default function PictureComponent({
  mobile,
  tablet,
  desktop,
  alt,
}: PictureProps) {
  return (
    <>
      <Image
        alt={alt}
        src={mobile}
        fill
        sizes="100vw"
        style={{ objectFit: "cover" }}
        className="block md:hidden"
      />
      <Image
        alt={alt}
        src={tablet}
        fill
        sizes="100vw"
        style={{ objectFit: "cover" }}
        className="hidden md:block lg:hidden"
      />
      <Image
        alt={alt}
        src={desktop}
        fill
        sizes="100vw"
        style={{ objectFit: "cover" }}
        className="hidden lg:block"
      />
    </>
  );
}
```

**Reply by bobaaaaa** [reply]
This will always load all three images no matter on which breakpoint you are. You can check the network panel in the dev tools

**Reply by kncalderong** [reply]
> This will always load all three images no matter on which breakpoint you are. You can check the network panel in the dev tools

No, it will not load all three images. It will load just the one depending on your screen size. 
![Screenshot 2025-06-04 092723](https://github.com/user-attachments/assets/6cc5bcc9-40a1-4868-8c3b-532465a17dec)

Test it...

**Reply by Multiply** [reply]
Try with `priority` on. :)
